lib/scholar/shared.ex

```diff
@@ -0,0 +1,28 @@
+defmodule Scholar.Shared do
+  @moduledoc false
+
+  # Collection of private helper functions and
+  # macros for enforcing shape/type constraints,
+  # and doing shape calculations.
+
+  import Nx.Defn
+
+  @doc """
+  Asserts `lhs` has same shape as `rhs`.
+  """
+  defn assert_shape!(lhs, rhs) do
```

Comment:
We already have `assert_shape` in Kernel, so my suggestion would be:

```suggestion
  defn assert_same_shape!(lhs, rhs) do
```